File: app/src/main/java/com/attendance/login/LoginSubscriber.java
package com.attendance.login;

import com.attendance.AttendanceApplication;
import com.attendance.R;

import java.net.ConnectException;
import java.net.SocketTimeoutException;

import rx.Subscriber;

/**
 * Created by peiqin on 3/5/2017.
 */

public class LoginSubscriber extends Subscriber<Boolean> {

    private LoginContract.View view;
    private LoginContract.Presenter presenter;

    public LoginSubscriber(LoginContract.Presenter presenter, LoginContract.View view) {
        this.view = view;
        this.presenter = presenter;
    }

    @Override
    public void onStart() {
        view.showProgress(AttendanceApplication.getContext().getString(R.string.login_loading));
    }

    @Override
    public void onCompleted() {
        view.cancelProgress();
        presenter.writeData();
        if (!this.isUnsubscribed()) {
            this.unsubscribe();
        }
    }

    @Override
    public void onError(Throwable e) {

        if (e instanceof SocketTimeoutException || e instanceof ConnectException) {
            view.showTip(AttendanceApplication.getContext().getString(R.string.network_corrupt));
        } else {
            view.showTip("error:" + e.getMessage());
        }
        view.cancelProgress();

        if (!this.isUnsubscribed()) {
            this.unsubscribe();
        }

    }

    @Override
    public void onNext(Boolean result) {

        if (result) {
            view.showTip(AttendanceApplication.getContext().getString(R.string.login_success));
            view.startMainActivity();
        } else {
            view.showTip(AttendanceApplication.getContext().getString(R.string.login_failed));
        }

    }

}

File: app/src/main/kotlin/com.attendance/login/LoginSubscriber.kt
package com.attendance.login

import com.attendance.AttendanceApplication
import com.attendance.R
import com.attendance.login.LoginContract

import java.net.ConnectException
import java.net.SocketTimeoutException

import rx.Subscriber

/**
 * Created by peiqin on 3/5/2017.
 */

class LoginSubscriber(private val presenter: LoginContract.Presenter, private val view: LoginContract.View) : Subscriber<Boolean>() {

    override fun onStart() {
        view.showProgress(AttendanceApplication.context.getString(R.string.login_loading))
    }

    override fun onCompleted() {
        view.cancelProgress()
        presenter.writeData()
        if (!this.isUnsubscribed) {
            this.unsubscribe()
        }
    }

    override fun onError(e: Throwable) {

        if (e is SocketTimeoutException || e is ConnectException) {
            view.showTip(AttendanceApplication.context.getString(R.string.network_corrupt))
        } else {
            view.showTip("error:" + e.message)
        }
        view.cancelProgress()

        if (!this.isUnsubscribed) {
            this.unsubscribe()
        }

    }

    override fun onNext(result: Boolean?) {

        if (result!!) {
            view.showTip(AttendanceApplication.context.getString(R.string.login_success))
            view.startMainActivity()
        } else {
            view.showTip(AttendanceApplication.context.getString(R.string.login_failed))
        }

    }

}
